---
title: "GWASr: Plotting GWAS Summary Statistics with GGPlot"
date: "12/13/2021"
output: html_document
author:
  - Franjo Ivankovic^[University of Florida, njofrica@ufl.edu]
  - Mingjing Chen^[University of Florida, mingjingchen@ufl.edu]
  - Xiaoxiao Wen^[University of Florida, xiaoxiao.wen@ufl.edu]
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  echo = TRUE
)
```

GWASr is a simple R pacakge which is capable of building QQ and Manhattan plots as well as calculating genomic lambda constant.

## Motivation
One of the most important, yet most difficult parts of complex statistical analyses of largee data sets, particularly in genomics, is visualising data. Exisiting functions that do so are split between many different packages, and usually involve a lot of function writing and re-writing to preprocess and actually plot the data.

We've created GWASr to circuimvent these issues and combin several important GWAS-related function into a single package. Moreover, both QQplot and Manhattan plots are returned as *ggplot* objects, which means that ideally the idividual should be able to furhter modify the plots aesthetically, without extenive coding - it's as simple as passing the entire dataframe into the function! 

Additionally, we include a C++ function to efficiently analyze millions of data points and calculate the genomic inflation factor, here also referred to as genomic lambda contant. This metric is a useful estimator of systemic biases in the GWAS model, and is widely used by the statitical genomics community.

## Install Packages
To install GWASr, a simple one-line code below with fetch this package from GitHub and install it in R. Additionally, the *build_vignettes = TRUE* option will automatically build an HTML of this vignette.
```{r install}
devtools::install_github("FranjoIM/GWASr", ref = "v0.1.0", build_vignettes = TRUE)
```

## Load the Package
Loading library is very easy, by simply running the following one-line code.
```{r setup}
library(GWASr)
```

## Accessing Vignettes
To access this vignette, simply run the following code in the R console.
```{r vignette}
vignette("GWASr", package = "GWASr")
```

## Included Datasets
This package includes two datasets which were obtained from [Psychiatric Genomics Consortium](https://www.med.unc.edu/pgc/download-results/) results page, and were originally published by [Demontis et al. (2019)](https://doi.org/10.1038/s41588-018-0269-7) in *Nature Genetics*.

### ADHD Meta
ADHD Meta is a processed data frame of 239,961 rows and 5 variables. In order to reduce memory requirements for packaging and distribution, as well as running trial functions, all observations with $p$ values over $0.2$ were excluded in the set.

The variables include:

1. CHR<chr>: chromosome ID
2. SNP<chr>: SNP ID
3. BP<num>: base-pair genomic location
4. OR<num>: Odds-ratio from GWAS
5. P<num>: P values from GWAS

As lazy-loading is on for this package, the data should load with the package and callable ass:
```{r ADHDMeta, eval = FALSE}
ADHDMeta

names(ADHDMeta)
#> [1] "CHR" "SNP" "BP"  "OR"  "P"

unique(ADHDMeta$CHR)
#> [1] "2"  "7"  "11" "8"  "9"  "17" "6"  "16" "1"  "4"  "15" "3"  "12" "5"  "13" "10" "18" "14" "19" "20" "21" "22"
```

### ADHD Meta P
ADHD Meta P is a minimally processed collection of SNP IDs and related $p$ values. It includes 8,094,094 rows and 2 variables. Most of the columns were removed in thee interest of reducing memory requirements, but all the $p$ values were kept as reported in the study.

The variables include:

1. SNP<chr>: SNP ID
2. P<num>: P values from GWAS

As lazy-loading is on for this package, the data should load with the package and callable ass:
```{r ADHDMetaP, eval = FALSE}
ADHDMetaP

names(ADHDMetaP)
#> [1] "SNP" "P"

min(ADHDMetaP$P)
#> [1] 6.452e-13

ADHDMetaP[which.min(ADHDMetaP$P),]
#>                SNP         P
#> 1103564 rs11420276 6.452e-13
```

## Overview of GWASr Functions
GWASr includes 3 functions, namely 2 plottng functions written in R (*plot_manhattan* and *plot_qq*), and 1 function written in Rcpp (*GLambda*). While the main focus of the package are the plotting functions, we will overview the *GLambda* first, as it's fairly straight forward. 

### Genomic Lambda Calculation (Rcpp)
Genomic lambda is a numerical constant, which is defined as the ratio of the median $\chi^2$ test score to an $\chi^2$ at $p=0.5$, or $0.4549364$. Genomic Lambda, also called the genomic inflation factor $\lambda_{gc}$, measures systematic biases that may be present in the GWAS association results. The expected $\lambda_{gc} \leq 1$. If the $\lambda_{gc} > 1$, then this may be evidence for some systematic bias that needs to be corrected in the analysis. 

The easiest way to calculate $\lambda_{gc}$ is to pass the p values from the summary statistics directly into the function. And example using the included data is:
```{r Genomic lambda}
GLambda(ADHDMetaP$P)
#> [1] 1.219852
```

Any numeric variable can be passed into the function. Additionally, by the very mathematical definition provided above, the $\lambda_{gc}$ when a single $p = 0.5$ is passed in should evaluate to $1$.
```{r Genomic lambda tes}
GLambda(0.5)
#> 1
```

For a detailed overview of arguments, run:
```{r Geenomic lambda arguments}
?GLambda
```

### Manhattan Plot
Manhattan plot is the most frequent method of visually representing the strength of genome-wide associations, by plotting each SNP's association $p$ on a y-axis, specifically as a $-log_{10}$ transform thereof, while plotting genomic positions on the x-axis. Genomic positions are mapped by typical *chr: base* address. Additionally, this function plots two reference lines on the plot at $p_{stringent} = 3*10^-8$ and $p_{lax} = 10^-5$, which represent genome-wide significant and genome-wide suggestive thresholds, respectively.

This function requires a data frame with 3 columns: chromosome ID (character class), base-pair position (numeric class), and $p$ values (numeric types). 
```{r Manhattan plot}
plot_manhattan(ADHDMeta)
```

For a detailed overview of arguments, run:
```{r Manhattan plot arguments}
?plot_manhattan
```

For the rest of this exercise, we will use a reduced data set for a more memory effcient experience. So let's do that here. We will randomly sample 10% of the included data set. Let's also assume that our file didn't arrive in a clean form, so let's rename some of the variables. Let's also change the chromosome data type from *character* to *numeric*.
```{r Data Reduction 1}
ADHDsub <- ADHDMeta[sample(1:nrow(ADHDMeta), 0.10*nrow(ADHDMeta)),]

names(ADHDsub)[names(ADHDsub) == 'CHR'] <- "chromosome"
names(ADHDsub)[names(ADHDsub) == 'P'] <- "p.value"
names(ADHDsub)[names(ADHDsub) == 'BP'] <- "basepair"

ADHDsub$chromosome <- as.numeric(ADHDsub$chromosome)
```

Now let's play with modifying different arguments of the data frame. 

1. We need to make sure to specify the changed column names. 
2. Our function can handle chromosome as *numeric* so we don't have to worry about it. 
3. Let's set the more stringent form of the p-value threshold at $10^{-12}$ and make it yellow. 
4. Let's change the lax p-value threshold at $5*10^{-4}$ and make it green.
5. Let's change colors of the points to be #FDB4FF, #FFB4B4, and #FFDBB4.
6. Change plot name to "Funky Manhattan"

```{r manhattan 2}
plot_manhattan(data = ADHDsub, P = "p.value", CHR = "chromosome", BP = "basepair",
               sigA = 1e-12, sigAcol = "yellow", sigB = 5e-4, sigBcol = "green",
               colrs = c("#FDB4FF","#FFB4B4","#FFDBB4"), main = "Funky Manhattan")
```

### QQ Plot
In a genome-wide association study, most of the SNPs tested should not to be associated with the phenotype. Thus, ideally, qq-plot of GWAS $p$ should follow a straight diagonal line with expected $p$, only taperig off at the very end. QQ plots show the quantile distribution of the $p$ values from our samples. QQ plots can inform whether there is a presence of systematic bias, much like the genomic lambda, and indicate if a closer look at data and model are necessary.

This function requres a data frame with 1 column: $p$ values (numeric types).
```{r QQ plot}
plot_qq(ADHDMetaP)
```

For a detailed overview of arguments, run:
```{r QQ plot arguments}
?plot_qq
```

Let's try plotting a simulated set of $p$ values. However, this time let's hide the reported genomic constant and SNP count. Additionally, let's make the line yellow, and points pink.
```{r qqplot 2}
plot_qq(data.frame(SimulP = ppoints(90000)), P = "SimulP",
        showN = FALSE, showGL = FALSE, lcol = "yellow", pcol = "pink")
```
And that's it! Make sure to check out the function descriptions and enjoy plaing with it! Issues can be reported via GitHub in the [GWASr > Issues](https://github.com/FranjoIM/GWASr/issues).

## References
* Demontis, D., Walters, R.K., Martin, J. et al. Discovery of the first genome-wide significant risk loci for attention deficit/hyperactivity disorder. *Nat Genet* 51, 63â€“75 (2019). [10.1038/s41588-018-0269-7](https://doi.org/10.1038/s41588-018-0269-7)
